dev-docs/plugin-dev-notes/README.md

```diff
@@ -0,0 +1,106 @@
+
+# Plugin dev notes
+
+Notes on creating plugins and extensions for mewbot.
+
+## Plugin writing intro
+
+### Why you might want to write a plugin?
+
+There are two reasons - and two types of plugin you might want to write.
+
+#### 1) Extending mewbot itself
+
+Base mewbot is more a framework than a complete program.
+It's designed to take yaml files, load them, and turn them into bots which can interact with the world.
+While it does have some basic capabilities, the majority of its capabilities come from plugins
+(either plugins already bundled with the program or plugins written for it.)
+
+Thus, if you want to extend the capabilities of the _framework itself_ you probably want to write a **mewbot framework plugin**.
+
+On a technical level this will take the form of a namespace plugin which will use the `mewbot_framework_plugin_template`.
+
+E.g. I would like mewbot to be able to talk to "Insert communications protocol here".
+I will write a plugin which will present a new IOConfig in `mewbot.io`
+
+For further information on how to write namespace plugins, please see the python core documentation - [python namespace plugin docs](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/).
+
+#### 2) Creating tools for bots
+
+If, on the other hand, you want to write some tools which bots _themselves_ will use, then you probably want to write a **mewbot bot plugin**.
+
+This would be a plugin for mewbot which does not add to the core module namespace.
+It might include detailed Triggers, Conditions and Actions that you will use to build your bot via yaml.
+
+It might even include custom IOConfigs.
+Though, if you're going to the effort of writing one, please consider either
+1) extending an existing IOConfig with new capabilities
+2) writing a framework plugin to expose your IOConfig more broadly.
+
+### Which one should I use?
+
+As a rule of thumb
+1) If the plugin components should appear in the core namespace, write a **framework plugin**
+2) If the plugin components are to support a specific bot or a collection of bots, then write a **bot plugin**
+
+E.g.
+
+#### Example 1
+
+"I want to allow mewbot to talk to mastadon"
+
+Probably write a **framework** plugin.
+
+In fact, please do!
+That would really be very helpful.
+
+Your generic IOConfig will take the form of a namespace plugin.
+It will be used by importing it using something like
+
+```python
+from mewbot.io.mastadon import MastadonIOConfig
+```
```

Comment:
```suggestion
"I want to allow mewbot to talk to mastodon"

Probably write a **framework** plugin.

In fact, please do!
That would really be very helpful.

Your generic IOConfig will take the form of a namespace plugin.
It will be used by importing it using something like

```python
from mewbot.io.mastodon import MastodonIOConfig
```
```

Reply:
I've gone through and re-written this section to be more definite "probably write" has become "write" - in line with the suggestion below.